### PR TITLE
Produce expansion info for more builtin macros

### LINKED ISSUE #43268: `while cfg!(unix) { ... }` suggests to use `loop` instead
If any linting or suggestion is done in this case, it should suggest `if cfg!(unix) { loop { ... } }`

cc @jseyfried this has been in rustc forever, but clippy is only lately going crazy on it. Could this have come from your PR, too?

### PR DESCRIPTION
**Pull request by oli-obk**
r? @jseyfried 

fixes #43268

### CODE REVIEW

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -75,9 +75,11 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for WhileTrue {
         if let hir::ExprWhile(ref cond, ..) = e.node {
             if let hir::ExprLit(ref lit) = cond.node {
                 if let ast::LitKind::Bool(true) = lit.node {
-                    cx.span_lint(WHILE_TRUE,
-                                 e.span,
-                                 "denote infinite loops with loop { ... }");
+                    if lit.span.ctxt().outer().expn_info().is_none() {
```
Comment (jseyfried) [inline src/librustc_lint/builtin.rs]: I believe this could just be `if lit.span.ctxt() == SyntaxContext::empty()`.

**Comment by jseyfried** [comment]
Thanks! r=me with comment addressed.

**Comment by kennytm** [comment]
@bors retry #44221 (note: 2nd try)

<details>

```
[0KTesting librustc stage1 (i686-apple-darwin -> i686-apple-darwin)
[02:13:42] [m[m[33m[1mwarning:[m a `-j` argument was passed to Cargo but Cargo is also configured with an external jobserver in its environment, ignoring the `-j` parameter
[02:13:43] [m[m[32m[1m   Compiling[m fmt_macros v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/libfmt_macros)
[02:13:46] [m[m[32m[1m   Compiling[m arena v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/libarena)
[02:13:49] [m[m[32m[1m   Compiling[m serialize v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/libserialize)
[02:14:31] [m[m[32m[1m   Compiling[m graphviz v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/libgraphviz)
[02:14:35] [m[m[32m[1m   Compiling[m rustc_bitflags v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_bitflags)
[02:14:36] [m[m[32m[1m   Compiling[m rustc_platform_intrinsics v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_platform_intrinsics)
[02:14:42] [m[m[32m[1m   Compiling[m rustc_apfloat v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_apfloat)
[02:14:44] [m[m[32m[1m   Compiling[m rustc_data_structures v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_data_structures)
[02:15:43] [m[m[32m[1m   Compiling[m syntax_pos v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/libsyntax_pos)
[02:15:47] [m[m[32m[1m   Compiling[m rustc_errors v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_errors)
[02:15:52] [m[m[32m[1m   Compiling[m syntax v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/libsyntax)
[02:17:49] [m[m[32m[1m   Compiling[m rustc_const_math v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_const_math)
[02:17:52] [m[m[32m[1m   Compiling[m proc_macro v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/libproc_macro)
[02:17:55] [m[m[32m[1m   Compiling[m rustc_back v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_back)
[02:18:17] [m[m[32m[1m   Compiling[m rustc v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc)

No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

</details>

**Pull request merged**